activerecord/lib/active_record/relation/finder_methods.rb

```diff
@@ -147,9 +147,18 @@ def first!
     def last(limit = nil)
       return find_last(limit) if loaded?
 
-      result = order_values.empty? && primary_key ? order(arel_table[primary_key].desc) : reverse_order
-      result = result.limit!(limit || 1)
-      limit ? result.reverse : result.first
+      initial_limit = limit_value
+      sql_limit = [limit || 1, initial_limit || Float::INFINITY].min
+
+      result = limit(sql_limit)
+      if initial_limit
+        result.offset!((offset_value || 0) + initial_limit - sql_limit)
```

Comment:
The problem with behaviour comes from lack of tests. 

I tried some combinations of `offset` and `limit` on a relation where `last` is called and didn't find a different behaviour from previous version. Can you guess what else it could be? 

It would be cool to add a test anyway even we go with `to_a.last` when limit present on basic relation at the end anyway.